Import multiple envs when importing OpenAPI3

sixdouglas opened the pull request:
changelog(Improvements): Each OpenAPI server is now imported as a sub-environment

When the OpenApi file to import contains multiple servers, import each one as an environment.

Closes #3627

douglas-six committed:
feat(import): import multiple envs when importing OpenAPI3

When the OpenApi file to import contains multiple servers, import each one as an environment.

Closes #3627

ryanholden8 commented:
Really looking forward to using this, thank you!

develohpanda commented on packages/insomnia-importers/src/importers/openapi-3.ts:
```diff
@@ -110,6 +107,43 @@ const parseDocument = (rawData: string): OpenAPIV3.Document | null => {
+        const basePath = (currentServerUrl.pathname || '').replace(/\/$/, '');
+        const openapiEnv: ImportRequest = {
+          _type: 'environment',
+          _id: 'env___BASE_ENVIRONMENT_ID___sub',
```
This will cause all of the sub environments to have the same id and insomnia won't see them as unique.

In other areas we append a hash of the dependent property to make the id deterministic and unique. What would be a dependent property here - the URL? Ie. if the server URL changes, a _new_ sub environment will be created.

sixdouglas replied:
Nice catch. I'll fix it.

develohpanda commented on packages/insomnia-importers/src/importers/openapi-3.ts:
```diff
@@ -110,6 +107,43 @@ const parseDocument = (rawData: string): OpenAPIV3.Document | null => {
+          _type: 'environment',
+          _id: 'env___BASE_ENVIRONMENT_ID___sub',
+          parentId: baseEnv._id,
+          name: 'OpenAPI env',
```
Same note as above, the name for each sub environment will be `OpenAPI env`. What would you expect the environment created from each new server to be named?

sixdouglas replied:
I'll fix it too

develohpanda commented on packages/insomnia-importers/src/convert.test.ts:
```diff
@@ -10,3 +10,92 @@ describe('Import errors', () => {
   });
 });
+
+describe('Import OpenApi', () => {
```
Any particular reason you chose to add tests like this instead of adding to the fixtures in packages/insomnia-importers/src/importers/fixtures/openapi3?

sixdouglas replied:
No particular reason, just didn't understand the test mechanism. So I wrote this one.
I'll do it in the same way as the other tests.

develohpanda commented:
Thank you for the PR! A couple of notes, namely the `_id` and `name` generation for the sub-environments.

develohpanda reviewed (changes requested):
Thank you for the PR! A couple of notes, namely the `_id` and `name` generation for the sub-environments.

(The CI errors are linting, which should be fixed up with `npm run lint:fix`).

douglas-six committed:
feat(import): import multiple envs when importing OpenAPI3

When the OpenApi file to import contains multiple servers, import each one as an environment.

Closes #3627

antonioturdo commented:
I also hope to have this improvement in Insomnia.

dgmora commented:
This feels very necessary to work with insomnia when our primary source is OpenAPI. Is there a problem with the PR?

filfreire committed:
fix lint and tests & rebase with latest develop

filfreire reviewed (approved):
Sorry for the long wait @sixdouglas

Rebased with latest develop and fixed lint and tests.

Tested locally, LGTM! Waiting for someone else from @Kong/team-insomnia to give the 👍
